### DISCUSSION TITLE
XMLSec 1.3.0-rc1

### DISCUSSION THREAD

**Discussion by lsh123** [lsh123/xmlsec#593] [Announcements]
The XMLSec 1.3.0 release includes a large number of changes including several API / ABI breaking changes (hence version bump). Please test the [release candidate](https://www.aleksey.com/xmlsec/download/xmlsec1-1.3.0-rc1.tar.gz) and let me know if you run into any issues! I plan to have at least one more release candidate in 2-3 weeks before the release mid-April.

### core xmlsec and all xmlsec-crypto libraries:
* **(ABI breaking change)** Added support for the [KeyInfoReference Element](https://www.w3.org/TR/xmldsig-core1/#sec-KeyInfoReference).
* **(ABI breaking change)** Switched xmlSecSize to use size_t by default. Use "--enable-size-t=no" configure option ("size_t=no" on Windows) to restore the old behaviour (note that support for xmlSecSize being different from size_t will be removed in the future).
* **(API breaking change)** Changed the key search to strict mode: only keys referenced by KeyInfo are used. To restore the old "lax" mode, set XMLSEC_KEYINFO_FLAGS_LAX_KEY_SEARCH flag on xmlSecKeyInfoCtx or use '--lax-key-search' option for XMLSec command line utility.
* **(API breaking change)** The KeyName element content is now trimmed before key search is performed.
* **(API breaking change)** Disabled FTP support by default. Use "--enable-ftp" configure option to restore it. Also added "--enable-http" and "--enable-files" configure options to control support for loading files over HTTP or locally.
* **(API/ABI breaking change)** Disabled MD5 digest method by default. Use "--enable-md5" configure options ("legacy-crypto" option on Windows) to re-enable MD5.
* **(ABI breaking change)** Removed deprecated functions.
* Added support for loading keys through [ossl-store](https://www.openssl.org/docs/man3.0/man7/ossl_store.html) interface (e.g. for using keys from an HSM). Also see '--privkey-openssl-store' and '--pubkey-openssl-store ' command line options for XMLSec utility.
* Added ability to control transforms binary chunk size to improve performance (see '--transform-binary-chunk-size' command line option for XMLSec utility).
* Fixed all potentially unsafe integer conversions and all the other warnings.
* Added [XML Signature 1.1 interop (2012)](https://www.w3.org/TR/2012/NOTE-xmldsig-core1-interop-20121113/) and [XML Encryption 1.1 interop (2012)](https://www.w3.org/TR/2012/NOTE-xmlenc-core1-interop-20121113/) tests.

### xmlsec-openssl library:
* Added support for [SHA3 digests](https://www.ietf.org/rfc/rfc9231.html#name-sha-3-algorithms).
* Added support for [ECDSA-SHA3 signatures](https://www.ietf.org/rfc/rfc9231.html#name-ecdsa-sha-ecdsa-ripemd160-e).
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [ConcatKDF key](https://www.w3.org/TR/xmlenc-core1/#sec-ConcatKDF) and [PBKDF2](https://www.w3.org/TR/xmlenc-core1/#sec-PBKDF2) derivation algorithms.
* **(ABI breaking change)** Added support for [ECDH-ES Key Agreement algorithm](https://www.w3.org/TR/xmlenc-core1/#sec-ECDH-ES).
* **(ABI breaking change)** Added support for [DH-ES Key Agreement algorithm](https://www.w3.org/TR/xmlenc-core1/#sec-DHKeyAgreementExplicitKDF) with explicit KDF.
* Added support for [MGF1 algorithm to RSA OAEP key transport](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).
* Added support for [X509Digest](https://www.w3.org/TR/xmldsig-core1/#sec-X509Data) element and ability to lookup keys using other X509Data elements.
* Added support for [DEREncodedKeyValue](https://www.w3.org/TR/xmldsig-core1/#sec-DEREncodedKeyValue) element.
* Automatically set key name from PKCS12 key name.
* Removed support for OpenSSL 1.0.0 and LibreSSL before 2.7.0.

### xmlsec-nss library:
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [RSA OAEP key transport including MGF1 algorithms](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).
* Added support for [AES GCM ciphers](https://www.w3.org/TR/xmlenc-core1/#sec-AES-GCM).
* Added support for [PBKDF2](https://www.w3.org/TR/xmlenc-core1/#sec-PBKDF2) derivation algorithms.
* Added support for [X509Digest](https://www.w3.org/TR/xmldsig-core1/#sec-X509Data) element and ability to lookup keys using other X509Data elements.
* Added support for [DEREncodedKeyValue](https://www.w3.org/TR/xmldsig-core1/#sec-DEREncodedKeyValue) element.
* Automatically set key name from PKCS12 key name.

### xmlsec-gnutls library:
* **(API/ABI breaking change)** Removed dependency on xmlsec-gcrypt and libgcrypt libraries (including API functions) to enable support for different GnuTLS backends.
* Bumped minimal GnuTLS version to 3.6.13.
* Added support for [SHA3 digests](https://www.ietf.org/rfc/rfc9231.html#name-sha-3-algorithms).
* Added support for [ECDSA signatures](https://www.w3.org/TR/xmldsig-core1/#sec-ECDSA).
* Added support for [DSA-SHA256 signatures](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [RSA PKCS 1.5 key transport](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-1_5).
* Added support for [AES GCM ciphers](https://www.w3.org/TR/xmlenc-core1/#sec-AES-GCM).
* Added support for [PBKDF2](https://www.w3.org/TR/xmlenc-core1/#sec-PBKDF2) derivation algorithms.
* Added support for [X509Digest](https://www.w3.org/TR/xmldsig-core1/#sec-X509Data) element and ability to lookup keys using other X509Data elements.
* Added support for [DEREncodedKeyValue](https://www.w3.org/TR/xmldsig-core1/#sec-DEREncodedKeyValue) element.
* Automatically set key name from PKCS12 key name.

### xmlsec-mscng library:
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [MGF1 algorithm to RSA OAEP key transport](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).
* **(ABI breaking change)** Added support for [ECDH-ES Key Agreement algorithm](https://www.w3.org/TR/xmlenc-core1/#sec-ECDH-ES).
* Added support for [ConcatKDF key](https://www.w3.org/TR/xmlenc-core1/#sec-ConcatKDF) and [PBKDF2](https://www.w3.org/TR/xmlenc-core1/#sec-PBKDF2) derivation algorithms.
* Added support for [X509Digest](https://www.w3.org/TR/xmldsig-core1/#sec-X509Data) element for keys and certificates lookup from the system stores (only SHA1 is supported).
* Added support for [DEREncodedKeyValue](https://www.w3.org/TR/xmldsig-core1/#sec-DEREncodedKeyValue) element.
* Automatically set key name from PKCS12 key name.

### xmlsec-mscrypto library:
* In maintenance mode starting from this release.
* Disabled by default support for NT4. Use "nt4=yes" configure option on Windows to re-enable it.

### xmlsec-gcrypt library:
* In maintenance mode starting from this release.
* Added support for [SHA3 digests](https://www.ietf.org/rfc/rfc9231.html#name-sha-3-algorithms).
* Added support for [ECDSA signatures](https://www.w3.org/TR/xmldsig-core1/#sec-ECDSA).
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [RSA PKCS 1.5 key transport](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-1_5).
* Added support for [RSA OAEP key transport including MGF1 algorithms](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).

### xmlsec command line utility:
* **(API breaking change)** The XMLSec command line utility is using 'strict' key search mode by default. To restore the old 'lax' key search mode, use the new '--lax-key-search' option.
* Added '--transform-binary-chunk-size' option to control transforms binary chunk size (increasing the chunk size should improve performance at the expense of memory usage.
* Added support for loading keys through [ossl-store](https://www.openssl.org/docs/man3.0/man7/ossl_store.html) interface (e.g. for using keys from an HSM). Also see '--privkey-openssl-store' and '--pubkey-openssl-store ' command line options for XMLSec utility.
* Added '--enabled-key-info-reference-uris' option to control processing of the the [KeyInfoReference Element](https://www.w3.org/TR/xmldsig-core1/#sec-KeyInfoReference).
* Added '--pbkdf2-key' option for loading PBKDF2 keys.
* Added '--concatkdf-key' option for loading ConcatKDF keys.
* Added '--hmac-min-out-len' option to control the min accepted HMAC Output length.
* Added '--pubkey-openssl-engine' option to load public keys from OpenSSL engine.
* Added '--crl-pem' and '--crl-der' options to load CRLs.
* Added '--verify-keys' option to verify key's certificate before loading into Keys Manager (only supported for OpenSSL currently).
* Enabled templatized output filenames to facilitate batch operations on multiple input files.

<hr /><em>This discussion was created from the release <a href='https://github.com/lsh123/xmlsec/releases/tag/xmlsec_1_3_0_rc1'>1.3.0-rc1</a>.</em>

**Comment by vmiklos** [comment]
The tarball at https://github.com/lsh123/xmlsec/releases/tag/xmlsec_1_3_0_rc1 is called  xmlsec1-1.2.30-rc1.tar.gz, I assume you wanted to call it xmlsec1-1.3.0-rc1.tar.gz?

**Reply by lsh123** [reply]
Argh, thanks -- fixed

**Comment by vmiklos** [comment]
Thanks! Hmm, something is still bad with the tarball, I think. I thought I would drop the tarball to a libreoffice build to see what happens, but the build breaks with:

```
  CC       membuf.lo
  CC       nodeset.lo
keysdata.c:34:10: fatal error: xmlsec/x509.h: No such file or directory
 #include <xmlsec/x509.h>
          ^~~~~~~~~~~~~~~
compilation terminated.
make[5]: *** [Makefile:697: keysdata.lo] Error 1
make[5]: *** Waiting for unfinished jobs....
In file included from kw_aes_des.c:28:0:
keysdata_helpers.h:21:10: fatal error: xmlsec/x509.h: No such file or directory
 #include <xmlsec/x509.h>
          ^~~~~~~~~~~~~~~
compilation terminated.
```

and indeed that header seems to be missing:

```
$ find workdir/UnpackedTarball/xmlsec -name x509.h
workdir/UnpackedTarball/xmlsec/include/xmlsec/mscng/x509.h
workdir/UnpackedTarball/xmlsec/include/xmlsec/nss/x509.h
workdir/UnpackedTarball/xmlsec/include/xmlsec/openssl/x509.h
workdir/UnpackedTarball/xmlsec/include/xmlsec/gnutls/x509.h
workdir/UnpackedTarball/xmlsec/include/xmlsec/mscrypto/x509.h
```

vs if I build from git, I have:

```
$ find . -name x509.h
./include/xmlsec/mscng/x509.h
./include/xmlsec/nss/x509.h
./include/xmlsec/openssl/x509.h
./include/xmlsec/x509.h
./include/xmlsec/gnutls/x509.h
./include/xmlsec/mscrypto/x509.h
```

Perhaps that file needs adding to the header list in include/xmlsec/Makefile.am?

**Reply by lsh123** [reply]
yep, this is why I do RC builds :( Thank you, RC2 coming up.

**Reply by lsh123** [reply]
[1.3.0-rc2](https://github.com/lsh123/xmlsec/releases/tag/xmlsec_1_3_0_rc2)

Thank you!
